File: pylzy/lzy/api/v2/remote_grpc/model/slot.py
import base64
from pathlib import Path
from typing import cast

import cloudpickle

from ai.lzy.v1.zygote_pb2 import _DATASCHEME  # type: ignore
from ai.lzy.v1.zygote_pb2 import _SLOT_DIRECTION  # type: ignore
from ai.lzy.v1.zygote_pb2 import _SLOT_MEDIA  # type: ignore
from ai.lzy.v1.zygote_pb2 import DataScheme, Slot
from lzy.api.v2.remote_grpc.model._pickle import pickle


def opposite(direction: _SLOT_DIRECTION):
    map_to_opposite = {
        _SLOT_DIRECTION.INPUT: _SLOT_DIRECTION.OUTPUT,
        _SLOT_DIRECTION.OUTPUT: _SLOT_DIRECTION.INPUT,
    }
    return map_to_opposite[direction]


def file_slot_t(
    name: Path,
    direction: _SLOT_DIRECTION,
    type_: type,
) -> Slot:
    return file_slot(name, direction, dump_type(type_))


def file_slot(
    name: Path,
    direction: _SLOT_DIRECTION,
    data_schema: _DATASCHEME,
) -> Slot:
    return Slot(
        name=str(name),
        media=_SLOT_MEDIA.FILE,
        direction=direction,
        contentType=data_schema,
    )


def dump_type(type_: type) -> DataScheme:
    return DataScheme(
        type=pickle(type_),
        schemeType=_DATASCHEME.cloudpickle,
    )

File: model/src/main/java/ai/lzy/model/slot/Slot.java
package ai.lzy.model.slot;

import ai.lzy.model.DataScheme;

public interface Slot {
    String STDOUT_SUFFIX = "stdout";
    String STDERR_SUFFIX = "stderr";

    Slot ARGS = new TextLinesInSlot("/dev/args");
    Slot STDIN = new TextLinesInSlot("/dev/stdin");
    Slot STDOUT = new TextLinesOutSlot("/dev/" + Slot.STDOUT_SUFFIX);
    Slot STDERR = new TextLinesOutSlot("/dev/" + Slot.STDERR_SUFFIX);

    String name();

    Media media();

    Direction direction();

    DataScheme contentType();

    enum Direction {
        INPUT,
        OUTPUT
    }

    enum Media {
        FILE(java.nio.file.Path.class),
        PIPE(java.nio.file.Path.class),
        ARG(java.lang.String.class);

        private final Class type;

        Media(Class type) {
            this.type = type;
        }

        public Class of() {
            return type;
        }
    }
}
